server/src/main/java/org/elasticsearch/ReleaseVersions.java

```diff
@@ -0,0 +1,128 @@
+
+    static {
+        USES_VERSIONS = ExtensionLoader.loadSingleton(ServiceLoader.load(BuildExtension.class))
+            .map(BuildExtension::hasReleaseVersioning)
+            .orElse(true);
+    }
+
+    private static final Map<Class<?>, NavigableMap<Integer, String>> RESOLVED_VERSIONS = new ConcurrentHashMap<>();
+
+    private static final Pattern COMMENT_LINE = Pattern.compile("^\\h*#");
+    private static final Pattern VERSION_LINE = Pattern.compile("(\\d+\\.\\d+\\.\\d+),(\\d+)(\\h*#.*)?");
+
+    static NavigableMap<Integer, String> readVersionsFile(Class<?> versionContainer) {
+        String versionsFileName = versionContainer.getSimpleName() + ".csv";
+        URL versionsFile = versionContainer.getResource(versionsFileName);
+        if (versionsFile == null) {
+            Log.error("Could not find versions file for class [{}]", versionContainer);
+            return Collections.emptyNavigableMap();
+        }
+
+        try (
+            BufferedReader reader = new BufferedReader(
+                new InputStreamReader(versionContainer.getResourceAsStream(versionsFileName), StandardCharsets.UTF_8)
+            )
+        ) {
+            NavigableMap<Integer, String> versions = new TreeMap<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (COMMENT_LINE.matcher(line).find()) continue; // a comment
+
+                var matcher = VERSION_LINE.matcher(line);
+                if (matcher.matches() == false) {
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile);
+                    continue;
+                }
+                try {
+                    Integer id = Integer.valueOf(matcher.group(2));
+                    String version = matcher.group(1);
+                    String existing = versions.putIfAbsent(id, version);
+                    if (existing != null) {
+                        Log.warn("Duplicate id [{}] for versions [{}, {}] in [{}]", id, existing, version, versionsFile);
+                    }
+                } catch (NumberFormatException e) {
+                    // cannot happen??? regex is wrong...
+                    assert false : "Regex allowed non-integer id through: " + e;
+                    Log.warn("Incorrect format for line [{}] in [{}]", line, versionsFile, e);
+                }
+            }
+
+            return Collections.unmodifiableNavigableMap(versions);
+        } catch (Exception e) {
+            Log.error("Could not read versions file [{}]", versionsFile, e);
+            return Collections.emptyNavigableMap();
+        }
+    }
+
+    private static NavigableMap<Integer, String> findVersionsMap(Class<?> versionContainer) {
+        return RESOLVED_VERSIONS.computeIfAbsent(versionContainer, ReleaseVersions::readVersionsFile);
+    }
+
+    public static String findReleaseVersion(Class<?> versionContainer, int id) {
+        if (USES_VERSIONS == false) return Integer.toString(id);
```

Comment:
I think this means in serverless we would print the raw id. Is that what we want? The id is just as useless to a user there. Are we assuming a serverless user should never hit an error that would need to print the release version since it is kept up to date?

Reply:
This is mostly aimed at log messages - where transport/index versions do have meaning. If we ever need to print out anything for serverless, it should be the raw id, not the release version.